sycl/doc/extensions/proposed/sycl_ext_oneapi_record_event.asciidoc

```diff
@@ -0,0 +1,219 @@
+finalized.
+*Shipping software products should not rely on APIs defined in this
+specification.*
+
+
+== Overview
+
+This extension provides the ability to time the execution of commands in a
+queue without enabling profiling on the entire queue.
+This is more efficient on some platforms because only a subset of the events
+are required to contain timestamp information.
+It is also more convenient for use in libraries, where the library wants to
+get timing information for some commands, but the library does not control the
+construction of the queue (which is where the `enable_profiling` property is
+passed).
+
+This extension is structured as a free function, rather than a member function
+on `queue`, in order to be consistent with the API design in
+link:../proposed/sycl_ext_oneapi_enqueue_functions.asciidoc[
+sycl_ext_oneapi_enqueue_functions]
+
+
+== Specification
+
+=== Feature test macro
+
+This extension provides a feature-test macro as described in the core SYCL
+specification.
+An implementation supporting this extension must predefine the macro
+`SYCL_EXT_ONEAPI_RECORD_EVENT` to one of the values defined in the table below.
+Applications can test for the existence of this macro to determine if the
+implementation supports this feature, or applications can test the macro's
+value to determine which of the extension's features the implementation
+supports.
+
+[%header,cols="1,5"]
+|===
+|Value
+|Description
+
+|1
+|The APIs of this experimental extension are not versioned, so the
+ feature-test macro always has this value.
+|===
+
+=== New device aspect
+
+This extension adds the `ext_oneapi_queue_event_recording` enumerator to the
+`sycl::aspect` enumeration.
+
+```
+namespace sycl {
+
+enum class aspect : /*unspecified*/ {
+  ext_oneapi_queue_event_recording
+};
+
+} // namespace sycl
+```
+
+When a device has this aspect, the `record_event` function may be called for a
+queue on this device even if the queue is not constructed with the property
+`property::queue::enable_profiling`.
+
+=== New free function
+
+This extension adds the following free function.
+
+|====
+a|
+[frame=all,grid=none]
+!====
+a!
+[source]
+----
+namespace sycl::ext::oneapi::experimental {
+
+event record_event(const queue& q);
+
+} // namespace sycl::ext::oneapi::experimental
+----
+!====
+
+_Effects:_ Enqueues a command barrier to `q`.
+Any commands submitted after this barrier cannot begin execution until all
+previously submitted commands have completed.
+
+_Returns:_ An event which represents the completion of the barrier.
+The event's status becomes `info::event_command_status::complete` when all
+commands submitted to the queue prior to the call to `record_event` have
+completed.
+The event's `info::event_profiling::command_submit` timestamp reflects the
+time at which `record_event` is called.
+The event's `info::event_profiling::command_end` timestamp reflects the time
+at which the event enters the "complete" state.
```

Comment:
> Unless `ext_oneapi_submit_barrier` extension is planned for retirement ...

I think you mean the [`sycl_ext_oneapi_enqueue_barrier`](https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/supported/sycl_ext_oneapi_enqueue_barrier.asciidoc) extension here.  Actually, we night be deemphasizing this extension in favor of the barrier functions in [`sycl_ext_oneapi_enqueue_functions`](https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/proposed/sycl_ext_oneapi_enqueue_functions.asciidoc).  However, I think your overall comment about using properties still applies.

In fact, we did consider exposing this functionality as a property instead of adding a new API.  In that alternative, we would allow the application to pass some property to either `ext_oneapi_submit_barrier` ([sycl_ext_oneapi_enqueue_barrier](https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/supported/sycl_ext_oneapi_enqueue_barrier.asciidoc)) or `submit_with_event` ([sycl_ext_oneapi_enqueue_functions](https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/proposed/sycl_ext_oneapi_enqueue_functions.asciidoc)), which enabled profiling timestamps in the returned event, even if the queue was not constructed with profiling enabled.

These were the arguments in favor of adding a new API like `record_event`:

* It wasn't clear that anyone really needed a general way to request profiling timing for events.  Rather, it seemed like the main request was for a way to insert timing markers before / after a sequence of commands.

* A new API like this is less verbose if you just want to time a sequence of commands in a queue.  If we added a property, you'd need to call either `ext_oneapi_submit_barrier` or `submit_with_event` and also pass the property value.

* The function name `record_event` seems more intuitive if you want to time commands.  It seems less intuitive to ask people to insert a barrier if their goal is to time commands.

* People migrating from CUDA will find the new API more familiar because its name and functionality more closely matches `cudaEventRecord`.

Reply:
> I think you mean the [sycl_ext_oneapi_enqueue_barrier](https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/supported/sycl_ext_oneapi_enqueue_barrier.asciidoc?rgh-link-date=2023-12-19T14%3A14%3A04Z) extension here.

Right.

> A new API like this is less verbose if you just want to time a sequence of commands in a queue. If we added a property, you'd need to call either ext_oneapi_submit_barrier or submit_with_event and also pass the property value.

Agree, but is this use case considered broad enough to warrant being less verbose? Making `device.get_info<sycl::info::device::name>()` or `sycl::property::queue::in_order()` less verbose will probably save much more keystrokes worldwide :)

We have such pattern in GROMACS, but that does not look like something used widely enough to warrant introducing a totally new concept (which takes mental space). And if someone wants to call it often, they are free to make a wrapper.

Also, if we want to be less verbose, perhaps "Throws: A synchronous `exception` with the `errc::invalid` error code if [...] the queue's device does not have the aspect `ext_oneapi_queue_event_recording`" should be reconsidered. Currently, if one wants to use this new functionality, every call to `record_event` will have to be wrapped in a conditional to check that the device supports it.

> The function name record_event seems more intuitive if you want to time commands. It seems less intuitive to ask people to insert a barrier if their goal is to time commands.

If we don't look at CUDA, I wouldn't call it intuitive. In other places, events are tightly associated with operations, so using the same name here already break some intuition about events one gleaned from the rest of the standard. Also, what exactly is "record"? Do we need to pause the recording later? Why recorded events always have timing enabled, but events from operation submission don't?

Barrier is much more apt, IMO: we request a barrier denoting the completion of all the previous operations; and it might have timing enabled if we want.

If "barrier" is no-go, I think `submit_marker` of `mark_event` are better names. "Marker", unlike "event", denotes something lacking a content, just a dot on a timeline. As a verb, "mark" clearly shows that this is a one-off action: event is marked, and that's it. I still think less is more, and since this thing already acts as a barrier, we might as well make it a barrier.

> People migrating from CUDA will find the new API more familiar because its name and functionality more closely matches cudaEventRecord.

The naming is already very different, and I find it strange to try to unify it here with CUDA while potentially making it less coherent with the rest of SYCL standard (graphs are recorded, barriers/operations are submitted, events are recorded, but in a different way than graphs).

As a person having migrated large codebase from CUDA, the indexing in multidimension `sycl::(nd_)item` was a much bigger pain then stream/queue, sub-group/warp, shift/shuffle etc naming.

Reply:
There has been an ongoing offline debate on the name of this API.  The name `submit_profiling_tag` has gotten some traction, so I updated the PR to show what this would look like (29bb178f97cde9e14f5bfe9d794b0c5112f520f8).

> If "barrier" is no-go, I think `submit_marker` of `mark_event` are better names. "Marker", unlike "event", denotes something lacking a content, just a dot on a timeline.

We liked this at first, but then we realized that the word "marker" means something different in [OpenCL](https://registry.khronos.org/OpenCL/specs/3.0-unified/html/OpenCL_API.html#clEnqueueMarker).  In order to avoid confusion, we substituted the name "profling_tag" for "marker".

> Also, if we want to be less verbose, perhaps "Throws: A synchronous `exception` with the `errc::invalid error` code if [...] the queue's device does not have the aspect `ext_oneapi_queue_event_recording`" should be reconsidered.

The problem is that not all backends can support this new API.  For example, OpenCL cannot return profiling information in an event unless the queue is constructed in a special way.  The only way we could support this extension on OpenCL would be to construct all OpenCL queues with profiling enabled, and we are unwilling to make that a requirement.  Therefore, we made this an optional feature, so that places the burden on applications to test whether the feature is supported.

Reply:
> The name submit_profiling_tag has gotten some traction

Agree, it's a good name.

> The problem is that not all backends can support this new API. For example, OpenCL cannot return profiling information in an event unless the queue is constructed in a special way. The only way we could support this extension on OpenCL would be to construct all OpenCL queues with profiling enabled, and we are unwilling to make that a requirement. Therefore, we made this an optional feature, so that places the burden on applications to test whether the feature is supported.

Yes, I understand that. My point was, there is little reason to value brevity of the function call ("If we added a property, you'd need to call either ext_oneapi_submit_barrier or submit_with_event and also pass the property value") when it, by design, has to be surrounded with extra boilerplate checking whether it is supported (not to mention `#ifdef`) or to have a complex logic around queue creation.